huaweicloud/services/iam/resource_huaweicloud_identity_virtual_mfa_device.go

```diff
@@ -0,0 +1,224 @@
+	qRCodePNG = strings.ReplaceAll(qRCodePNG, "{domain_name}", domainName.(string))
+	qRCodePNG = strings.ReplaceAll(qRCodePNG, "{user_name}", userName.(string))
+	qRCodePNG = strings.ReplaceAll(qRCodePNG, "{seed}", seed.(string))
+
+	d.Set("qr_code_png", qRCodePNG)
+
+	return nil
+}
+
+func resourceMFARead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	getMFAProduct := "iam"
+	getMFAClient, err := cfg.NewServiceClient(getMFAProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating IAM Client: %s", err)
+	}
+
+	getMFAHttpUrl := "v3.0/OS-MFA/users/{user_id}/virtual-mfa-device"
+	getMFAPath := getMFAClient.Endpoint + getMFAHttpUrl
+	getMFAPath = strings.ReplaceAll(getMFAPath, "{user_id}", d.Get("user_id").(string))
+	getMFAOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+	getMFAResp, err := getMFAClient.Request("GET", getMFAPath, &getMFAOpt)
+	if err != nil {
+		return diag.Errorf("error getting IAM virtual MFA device: %s", err)
+	}
+	getMFARespBody, err := utils.FlattenResponse(getMFAResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := jmespath.Search("virtual_mfa_device.serial_number", getMFARespBody)
+	if err != nil {
+		return diag.Errorf("error getting IAM virtual MFA device: serial_number is not found in API response")
+	}
+	d.SetId(id.(string))
+
+	index := strings.LastIndex(id.(string), ":mfa/") + len(":mfa/")
+	d.Set("name", id.(string)[index:])
+
+	return nil
+}
+
+func resourceMFADelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	dproduct := "iam"
+	client, err := cfg.NewServiceClient(dproduct, region)
+	if err != nil {
+		return diag.Errorf("error creating IAM Client: %s", err)
+	}
+
+	deleteMFAHttpUrl := "v3.0/OS-MFA/virtual-mfa-devices?user_id={user_id}&serial_number={id}"
+	deleteMFAPath := client.Endpoint + deleteMFAHttpUrl
+	deleteMFAPath = strings.ReplaceAll(deleteMFAPath, "{user_id}", d.Get("user_id").(string))
+	deleteMFAPath = strings.ReplaceAll(deleteMFAPath, "{id}", d.Id())
+	deleteMFAOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			204,
+		},
+	}
+	_, err = client.Request("DELETE", deleteMFAPath, &deleteMFAOpt)
+	if err != nil {
+		return diag.Errorf("error deleting IAM virtual MFA device: %s", err)
+	}
+
+	// When input any serial number, API always return 200 and no more information, so need to check it truly deleted
+	getMFAHttpUrl := "v3.0/OS-MFA/users/{user_id}/virtual-mfa-device"
+	getMFAPath := client.Endpoint + getMFAHttpUrl
+	getMFAPath = strings.ReplaceAll(getMFAPath, "{user_id}", d.Get("user_id").(string))
+	getMFAOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+	_, err = client.Request("GET", getMFAPath, &getMFAOpt)
+	if err == nil {
+		return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{}, "error deleting IAM virtual MFA device")
```

Comment:
if we use `CheckDeletedDiag` and the status code is 404, it will not return the error message.

```suggestion
		return diag.Errorf("error deleting IAM virtual MFA device: the virtual MFA device still exists")
```